File: Cpp/AdventOfCode/src/2015/d11_NextPass.cpp
#pragma once

#include "2015/d11_NextPass.h"
SOLUTION(2015, 11) {
    constexpr std::string Increment(const std::string & input) {
        std::string result = input;
        for (s32 i = static_cast<s32>(result.size()) - 1; i >= 0; i--) {
            if (result[i] == 'z') {
                result[i] = 'a';
            }
            else {
                result[i]++;
                break;
            }
        }

        return result;
    }

    static_assert(Increment("a") == "b");
    static_assert(Increment("z") == "a");
    static_assert(Increment("azzz") == "baaa");
    static_assert(Increment("xyz") == "xza");

    constexpr std::string ReplaceInvalid(const std::string & input) {
        std::string result = input;
        size_t index = 0;
        bool found = false;
        for (; index < result.size(); index++) {
            if (result[index] == 'i') {
                result[index] = 'j';
                found = true;
                break;
            }

            if (result[index] == 'o') {
                result[index] = 'p';
                found = true;
                break;
            }
            if (result[index] == 'l') {
                result[index] = 'm';
                found = true;
                break;
            }
        }

        if (found) {
            index++;
            for (; index < result.size(); index++) {
                result[index] = 'a';
            }
        }

        return result;
    }


    constexpr bool ContainsStraight(const std::string & input) {
        for (auto i = 2; i < input.size(); i++) {
            if (input[i] == input[i - 1] + 1 && input[i] == input[i - 2] + 2) {
                return true;
            }
        }

        return false;
    }

    constexpr bool ContainsTwoPairs(const std::string & input) {
        bool onePair = false;
        for (auto i = 1; i < input.size(); i++) {
            if (input[i] == input[i - 1]) {
                if (onePair) {
                    return true;
                }
                else {
                    onePair = true;
                    i++;
                }
            }
        }

        return false;
    }

    constexpr bool IsValid(const std::string & input) {
        return ContainsStraight(input) && ContainsTwoPairs(input);
    }

    constexpr std::string NextPass(std::string previous) {
        std::string result = previous;
        do {
            result = Increment(result);
            result = ReplaceInvalid(result);
        } while (!IsValid(result));

        return result;
    }

    PART_ONE() {
        return NextPass(lines[0]);
    }
    PART_TWO() {
        return NextPass(NextPass(lines[0]));
    }

    TESTS() {
        static_assert(ReplaceInvalid("oil") == "paa");
        static_assert(ReplaceInvalid("happy") == "happy");
        
        static_assert(ContainsStraight("abc") == true);
        static_assert(ContainsStraight("aabbc") == false);
        static_assert(ContainsStraight("aabcc") == true);
        static_assert(ContainsStraight("cba") == false);
        static_assert(ContainsStraight("aaa") == false);

        static_assert(ContainsTwoPairs("aabb") == true);
        static_assert(ContainsTwoPairs("abcd") == false);
        static_assert(ContainsTwoPairs("aaa") == false);
        static_assert(ContainsTwoPairs("aaaa") == true);

        static_assert(IsValid("abbceffg") == false);
        static_assert(IsValid("abcdffaa") == true);
        static_assert(IsValid("ghjaabcc") == true);

        //static_assert(NextPass("ghijklmn") == "ghjaabcc");
        //static_assert(NextPass("vzbxkghb") == "vzbxxyzz");
        //static_assert(NextPass("vzbxxyzz") == "vzcaabcc");
        return true;
    }
}

File: Rust/rust_aoc/src/year_2015/d11_next_pass.rs
pub fn part_one(lines: &Vec<String>) -> String {
    next_pass(&lines[0])
}

pub fn part_two(lines: &Vec<String>) -> String {
    next_pass(&next_pass(&lines[0]))
}

fn next_pass(previous: &str) -> String {
    let mut result = String::from(previous);
    loop {
        increment(&mut result);
        replace_invalid(&mut result);
        if is_valid(&result) {
            break;
        }
    }

    result
}

fn increment(input: &mut String) {
    unsafe {
        for c in input.as_bytes_mut().iter_mut().rev() {
            if *c == b'z' {
                *c = b'a';
            } else {
                *c += 1;
                break;
            }
        }
    }
}

fn replace_invalid(input: &mut String) {
    let mut clear_rest = false;

    unsafe {
        for c in input.as_bytes_mut().iter_mut() {
            if clear_rest {
                *c = b'a';
            }
            if *c == b'i' || *c == b'o' || *c == b'l' {
                *c += 1;
                clear_rest = true;
            }
        }
    }
}

fn is_valid(input: &str) -> bool {
    contains_straight(input) && contains_two_pairs(input)
}

fn contains_straight(input: &str) -> bool {
    input.as_bytes().windows(3).any(|a| a[0] == (a[1] - 1) && a[0] == (a[2] - 2))
}

fn contains_two_pairs(input: &str) -> bool {
    input.as_bytes().windows(5).any(|a| a[0] == a[1] && a[0] != a[2] && a[2] != a[3] && a[3] == a[4])
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn increment_tests() {
        let mut val = String::from("a");
        increment(&mut val);
        assert_eq!(val, "b");
        
        val = String::from("z");
        increment(&mut val);
        assert_eq!(val, "a");

        val = String::from("azzz");
        increment(&mut val);
        assert_eq!(val, "baaa");

        val = String::from("xyz");
        increment(&mut val);
        assert_eq!(val, "xza");
    }

    #[test]
    fn replace_invalid_tests() {
        let mut val = String::from("oil");
        replace_invalid(&mut val);
        assert_eq!(val, "paa");

        val = String::from("happy");
        replace_invalid(&mut val);
        assert_eq!(val, "happy");
    }

    #[test]
    fn contains_straight_tests() {
        assert!(contains_straight("abc"));
        assert!(!contains_straight("aabbc"));
        assert!(contains_straight("aabcc"));
        assert!(!contains_straight("cba"));
        assert!(!contains_straight("aaa"));
    }

    #[test]
    fn contains_two_pairs_tests() {
        assert!(contains_two_pairs("aacbb"));
        assert!(!contains_two_pairs("abcde"));
        assert!(!contains_two_pairs("aaa"));
        assert!(contains_two_pairs("aabaa"));
    }

    #[test]
    fn is_valid_tests() {
        assert!(!is_valid("abbceffg"));
        assert!(is_valid("abcffdaa"));
        assert!(is_valid("ghjaabcc"));
    }

    #[test]
    fn next_pass_tests() {
        assert_eq!(next_pass("ghijklmn"), "ghjaabcc");
        assert_eq!(next_pass("vzbxkghb"), "vzbxxyzz");
        assert_eq!(next_pass("vzbxxyzz"), "vzcaabcc");
    }
}